// @smartthings/cli/dist/commands/devices/rename.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const inquirer_1 = tslib_1.__importDefault(require("inquirer"));
const cli_lib_1 = require("@smartthings/cli-lib");
const devices_1 = require("../devices");
class DeviceRenameCommand extends cli_lib_1.APICommand {
    async run() {
        var _a;
        const { args, argv, flags } = this.parse(DeviceRenameCommand);
        await super.setup(args, argv, flags);
        const id = await devices_1.chooseDevice(this, args.id);
        const label = (_a = args.label) !== null && _a !== void 0 ? _a : (await inquirer_1.default.prompt({
            type: 'input',
            name: 'label',
            message: 'Enter new device label:',
        })).label;
        await cli_lib_1.outputItem(this, { buildTableOutput: data => devices_1.buildTableOutput(this.tableGenerator, data) }, () => this.client.devices.update(id, { label }));
    }
}
exports.default = DeviceRenameCommand;
DeviceRenameCommand.description = 'rename a device';
DeviceRenameCommand.flags = {
    ...cli_lib_1.APICommand.flags,
    ...cli_lib_1.outputItem.flags,
};
DeviceRenameCommand.args = [
    {
        name: 'id',
        description: 'the device id',
    },
    {
        name: 'label',
        description: 'the new device label',
    },
];
