// @smartthings/plugin-cli-edge/lib/commands/edge/channels/create.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const inquirer_1 = __importDefault(require("inquirer"));
const cli_lib_1 = require("@smartthings/cli-lib");
const edge_command_1 = require("../../../lib/edge-command");
const tableFieldDefinitions = ['channelId', 'name', 'description', 'type', 'termsOfServiceUrl',
    'createdDate', 'lastModifiedDate'];
class ChannelsCreateCommand extends edge_command_1.EdgeCommand {
    async getInputFromUser() {
        const name = (await inquirer_1.default.prompt({
            type: 'input',
            name: 'name',
            message: 'Channel name:',
            validate: input => input ? true : 'name is required',
        })).name;
        const description = (await inquirer_1.default.prompt({
            type: 'input',
            name: 'description',
            message: 'Channel description:',
            validate: input => input ? true : 'description is required',
        })).description;
        const termsOfServiceUrl = (await inquirer_1.default.prompt({
            type: 'input',
            name: 'termsOfServiceUrl',
            message: 'Channel terms of service URL:',
            validate: input => input ? true : 'termsOfServiceUrl is required',
        })).termsOfServiceUrl;
        return { name, description, termsOfServiceUrl, type: 'DRIVER' };
    }
    async run() {
        const { args, argv, flags } = this.parse(ChannelsCreateCommand);
        await super.setup(args, argv, flags);
        await cli_lib_1.inputAndOutputItem(this, { tableFieldDefinitions }, (_, input) => this.edgeClient.channels.create(input), cli_lib_1.userInputProcessor(this));
    }
}
exports.default = ChannelsCreateCommand;
ChannelsCreateCommand.description = 'create a channel';
ChannelsCreateCommand.flags = {
    ...edge_command_1.EdgeCommand.flags,
    ...cli_lib_1.inputAndOutputItem.flags,
};
