// @smartthings/plugin-cli-edge/lib/commands/edge/channels/enrollments.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cli_lib_1 = require("@smartthings/cli-lib");
const edge_command_1 = require("../../../lib/edge-command");
const install_1 = require("../drivers/install");
class ChannelsEnrollmentsCommand extends edge_command_1.EdgeCommand {
    async run() {
        const { args, argv, flags } = this.parse(ChannelsEnrollmentsCommand);
        await super.setup(args, argv, flags);
        const config = {
            primaryKeyName: 'channelId',
            sortKeyName: 'name',
            listTableFieldDefinitions: ['channelId', 'name', 'description', 'createdDate', 'lastModifiedDate', 'subscriptionUrl'],
        };
        const hubId = await install_1.chooseHub(this, 'Select a hub.', args.idOrIndex, { allowIndex: true });
        await cli_lib_1.outputList(this, config, () => this.edgeClient.hubs.enrolledChannels(hubId));
    }
}
exports.default = ChannelsEnrollmentsCommand;
ChannelsEnrollmentsCommand.description = 'list all channels a given hub is enrolled in';
ChannelsEnrollmentsCommand.flags = {
    ...edge_command_1.EdgeCommand.flags,
    ...cli_lib_1.outputList.flags,
};
ChannelsEnrollmentsCommand.args = [{
        name: 'idOrIndex',
        description: 'the hub id or number in list',
    }];
