// @smartthings/plugin-cli-edge/lib/lib/endpoints/hubs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HubsEndpoint = void 0;
const core_sdk_1 = require("@smartthings/core-sdk");
class HubsEndpoint extends core_sdk_1.Endpoint {
    constructor(config) {
        super(new core_sdk_1.EndpointClient('hubdevices', config));
    }
    /**
     * Install the specified drivers into the specified hub in the specified location.
     */
    async installDriver(driverId, hubId, channelId) {
        return this.client.put(`${hubId}/drivers/${driverId}`, { channelId });
    }
    async uninstallDriver(driverId, hubId) {
        return this.client.delete(`${hubId}/drivers/${driverId}`);
    }
    async listInstalled(hubId) {
        return this.client.get(`${hubId}/drivers`);
    }
    async getInstalled(hubId, driverId) {
        return this.client.get(`${hubId}/drivers/${driverId}`);
    }
    async enrolledChannels(hubId) {
        return this.client.get(`${hubId}/channels`, { channelType: 'DRIVERS' });
    }
}
exports.HubsEndpoint = HubsEndpoint;
