// @smartthings/plugin-cli-edge/lib/commands/edge/channels/drivers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cli_lib_1 = require("@smartthings/cli-lib");
const edge_command_1 = require("../../../lib/edge-command");
const channels_1 = require("../channels");
class ChannelsDriversCommand extends edge_command_1.EdgeCommand {
    async run() {
        const { args, argv, flags } = this.parse(ChannelsDriversCommand);
        await super.setup(args, argv, flags);
        const config = {
            primaryKeyName: 'channelId',
            sortKeyName: 'version',
            listTableFieldDefinitions: ['channelId', 'driverId', 'version', 'createdDate', 'lastModifiedDate'],
        };
        const channelId = await channels_1.chooseChannel(this, 'Select a channel.', args.idOrIndex, { allowIndex: true, includeReadOnly: true });
        await cli_lib_1.outputList(this, config, () => this.edgeClient.channels.listAssignedDrivers(channelId));
    }
}
exports.default = ChannelsDriversCommand;
ChannelsDriversCommand.description = 'list all drivers assigned to a given channel';
ChannelsDriversCommand.flags = {
    ...edge_command_1.EdgeCommand.flags,
    ...cli_lib_1.outputList.flags,
};
ChannelsDriversCommand.args = [{
        name: 'idOrIndex',
        description: 'the channel id or number in list',
    }];
ChannelsDriversCommand.aliases = ['edge:channels:assignments'];
