// @smartthings/cli/dist/commands/locations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chooseLocation = exports.tableFieldDefinitions = void 0;
const cli_lib_1 = require("@smartthings/cli-lib");
exports.tableFieldDefinitions = [
    'name', 'locationId', 'countryCode', 'timeZoneId', 'backgroundImage',
    'latitude', 'longitude', 'regionRadius', 'temperatureScale', 'locale',
];
async function chooseLocation(command, locationFromArg) {
    const config = {
        itemName: 'location',
        primaryKeyName: 'locationId',
        sortKeyName: 'name',
    };
    return cli_lib_1.selectFromList(command, config, locationFromArg, () => command.client.locations.list());
}
exports.chooseLocation = chooseLocation;
class LocationsCommand extends cli_lib_1.APICommand {
    async run() {
        const { args, argv, flags } = this.parse(LocationsCommand);
        await super.setup(args, argv, flags);
        const config = {
            primaryKeyName: 'locationId',
            sortKeyName: 'name',
            tableFieldDefinitions: exports.tableFieldDefinitions,
        };
        await cli_lib_1.outputListing(this, config, args.idOrIndex, () => this.client.locations.list(), id => this.client.locations.get(id));
    }
}
exports.default = LocationsCommand;
LocationsCommand.description = 'list locations or get information for a specific Location';
LocationsCommand.flags = {
    ...cli_lib_1.APICommand.flags,
    ...cli_lib_1.outputListing.flags,
};
LocationsCommand.args = [{
        name: 'idOrIndex',
        description: 'the location id or number in list',
    }];
