// @smartthings/cli/dist/commands/schema/regenerate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cli_lib_1 = require("@smartthings/cli-lib");
class SchemaAppRegenerateCommand extends cli_lib_1.APICommand {
    async run() {
        const { args, argv, flags } = this.parse(SchemaAppRegenerateCommand);
        await super.setup(args, argv, flags);
        const config = {
            primaryKeyName: 'endpointAppId',
            sortKeyName: 'appName',
        };
        const id = await cli_lib_1.selectFromList(this, config, args.id, async () => await this.client.schema.list(), 'Select a schema app to regenerate its clientId and clientSecret.');
        await cli_lib_1.outputItem(this, { tableFieldDefinitions: ['endpointAppId', 'stClientId', 'stClientSecret'] }, () => this.client.schema.regenerateOauth(id));
    }
}
exports.default = SchemaAppRegenerateCommand;
SchemaAppRegenerateCommand.description = 'Regenerate the clientId and clientSecret of the ST Schema connector. The previous values will be invalidated, which may affect existing installations.';
SchemaAppRegenerateCommand.flags = {
    ...cli_lib_1.APICommand.flags,
    ...cli_lib_1.outputItem.flags,
};
SchemaAppRegenerateCommand.args = [{
        name: 'id',
        description: 'schema app id',
    }];
