// @smartthings/cli/dist/commands/presentation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildTableOutput = void 0;
const cli_lib_1 = require("@smartthings/cli-lib");
function buildTableOutput(tableGenerator, presentation) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
    const basicInfo = tableGenerator.buildTableFromItem(presentation, [
        { prop: 'presentationId', label: 'Presentation ID' }, { prop: 'manufacturerName', label: 'Manufacturer Name' }, 'iconUrl',
    ]);
    let dashboardStates = 'No dashboard states';
    if (((_a = presentation.dashboard) === null || _a === void 0 ? void 0 : _a.states) && presentation.dashboard.states.length > 0) {
        const subTable = tableGenerator.newOutputTable({ head: ['Label', 'Alternatives', 'Group'] });
        for (const state of presentation.dashboard.states) {
            const alternatives = ((_b = state.alternatives) === null || _b === void 0 ? void 0 : _b.length)
                ? state.alternatives.length
                : 'none';
            subTable.push([
                state.label,
                alternatives,
                state.group ? state.group : '',
            ]);
        }
        dashboardStates = `Dashboard States\n${subTable.toString()}`;
    }
    function buildDisplayTypeTable(items) {
        const subTable = tableGenerator.newOutputTable({ head: ['Display Type'] });
        for (const item of items) {
            subTable.push([item.displayType]);
        }
        return subTable.toString();
    }
    function buildLabelDisplayTypeTable(items) {
        const subTable = tableGenerator.newOutputTable({ head: ['Label', 'Display Type'] });
        for (const item of items) {
            subTable.push([item.label, item.displayType]);
        }
        return subTable.toString();
    }
    let dashboardActions = 'No dashboard actions';
    if ((_d = (_c = presentation.dashboard) === null || _c === void 0 ? void 0 : _c.actions) === null || _d === void 0 ? void 0 : _d.length) {
        dashboardActions = `Dashboard Actions\n${buildDisplayTypeTable(presentation.dashboard.actions)}`;
    }
    let dashboardBasicPlus = 'No dashboard basic plus items';
    if ((_f = (_e = presentation.dashboard) === null || _e === void 0 ? void 0 : _e.basicPlus) === null || _f === void 0 ? void 0 : _f.length) {
        dashboardBasicPlus = `Dashboard Basic Plus\n${buildDisplayTypeTable(presentation.dashboard.basicPlus)}`;
    }
    let detailViews = 'No detail views';
    if ((_g = presentation.detailView) === null || _g === void 0 ? void 0 : _g.length) {
        const subTable = tableGenerator.newOutputTable({ head: ['Capability', 'Version', 'Component'] });
        for (const detailView of presentation.detailView) {
            subTable.push([
                detailView.capability,
                detailView.version ? detailView.version : 'none',
                detailView.component,
            ]);
        }
        detailViews = `Detail Views\n${subTable.toString()}`;
    }
    let automationConditions = 'No automation conditions';
    if ((_j = (_h = presentation.automation) === null || _h === void 0 ? void 0 : _h.conditions) === null || _j === void 0 ? void 0 : _j.length) {
        automationConditions = `Automation Conditions\n${buildLabelDisplayTypeTable(presentation.automation.conditions)}`;
    }
    let automationActions = 'No automation actions';
    if ((_l = (_k = presentation.automation) === null || _k === void 0 ? void 0 : _k.actions) === null || _l === void 0 ? void 0 : _l.length) {
        automationActions = `Automation Actions\n${buildLabelDisplayTypeTable(presentation.automation.actions)}`;
    }
    return `Basic Information\n${basicInfo}\n\n` +
        `${dashboardStates}\n\n` +
        `${dashboardActions}\n\n` +
        `${dashboardBasicPlus}\n\n` +
        `${detailViews}\n\n` +
        `${automationConditions}\n\n` +
        `${automationActions}\n\n` +
        '(Information is summarized, for full details use YAML, -y, or JSON flag, -j.)';
}
exports.buildTableOutput = buildTableOutput;
class PresentationCommand extends cli_lib_1.APICommand {
    async run() {
        const { args, argv, flags } = this.parse(PresentationCommand);
        await super.setup(args, argv, flags);
        const config = {
            buildTableOutput: (data) => buildTableOutput(this.tableGenerator, data),
        };
        await cli_lib_1.outputItem(this, config, () => this.client.presentation.getPresentation(args.presentationId, args.manufacturerName));
    }
}
exports.default = PresentationCommand;
PresentationCommand.description = 'query device presentation by vid';
PresentationCommand.flags = {
    ...cli_lib_1.APICommand.flags,
    ...cli_lib_1.outputItem.flags,
};
PresentationCommand.args = [
    {
        name: 'presentationId',
        description: 'system generated identifier that corresponds to a device presentation',
        required: true,
    },
    {
        name: 'manufacturerName',
        description: 'manufacturer name. Defaults to SmartThingsCommunity',
        required: false,
    },
];
PresentationCommand.examples = [
    '$ smartthings presentation fd4adb7f-4a23-4134-9b39-05ed889a03cf',
    '$ smartthings presentation 4ea31e30-2aba-41c7-a3ec-8f97423d565a SmartThings',
    '$ smartthings presentation fd4adb7f-4a23-4134-9b39-05ed889a03cf --language=ko',
    '$ smartthings presentation fd4adb7f-4a23-4134-9b39-05ed889a03cf --language=NONE',
    '',
    'Specifying only the presentationId defaults to the "SmartThingsCommunity" manufacturer',
    'name and the language set for the computer\'s operating system. The language can be',
    'overridden by specifying an ISO language code. If "NONE" is specified for the language',
    'flag then no language header is specified in the API request',
];
